drivers/espi/espi_handlers.c

```diff
@@ -0,0 +1,226 @@
+
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, config));
+	Z_OOPS(z_user_from_copy(&cfg_copy, cfg,
+				sizeof(struct espi_cfg)));
+
+	return z_impl_espi_config(dev, &cfg_copy);
+}
+#include <syscalls/espi_config_mrsh.c>
+
+static inline bool z_vrfy_espi_get_channel_status(struct device *dev,
+						  enum espi_channel ch)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, get_channel_status));
+
+	return z_impl_espi_get_channel_status(dev, ch);
+}
+#include <syscalls/espi_get_channel_status_mrsh.c>
+
+static inline int z_vrfy_espi_read_lpc_request(struct device *dev,
+					       enum lpc_peripheral_opcode op,
+					       u32_t *data)
+{
+	int ret;
+	u32_t data_copy;
+
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, read_lpc_request));
+	Z_OOPS(Z_SYSCALL_MEMORY_WRITE(data, sizeof(*data)));
+
+	ret = z_impl_espi_read_lpc_request(dev, op, &data_copy);
+
+	*data = data_copy;
+	return ret;
+}
+#include <syscalls/espi_read_lpc_request_mrsh.c>
+
+static inline int z_vrfy_espi_write_lpc_request(struct device *dev,
+						enum lpc_peripheral_opcode op,
+						u32_t *data)
+{
+	u32_t data_copy;
+
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, write_lpc_request));
+	Z_OOPS(Z_SYSCALL_MEMORY_READ(data, sizeof(*data)));
+
+	data_copy = *data;
+
+	return z_impl_espi_write_lpc_request(dev, op, data_copy);
+}
+#include <syscalls/espi_write_lpc_request_mrsh.c>
+
+static inline int z_vrfy_espi_send_vwire(struct device *dev,
+					 enum espi_vwire_signal signal,
+					 u8_t level)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, send_vwire));
+
+	return z_impl_espi_send_vwire(dev, signal, level);
+}
+#include <syscalls/espi_send_vwire_mrsh.c>
+
+static inline int z_vrfy_espi_receive_vwire(struct device *dev,
+					    enum espi_vwire_signal signal,
+					    u8_t *level)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, receive_vwire));
+	Z_OOPS(Z_SYSCALL_MEMORY_WRITE(level, sizeof(*level)));
+
+	return z_impl_espi_receive_vwire(dev, signal, level);
+}
+#include <syscalls/espi_receive_vwire_mrsh.c>
+
+static inline int z_vrfy_espi_read_request(struct device *dev,
+					   struct espi_request_packet *req)
+{
+	int ret;
+	struct  espi_request_packet req_copy;
+
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, read_request));
+	Z_OOPS(z_user_from_copy(&req_copy, req,
+				sizeof(struct espi_request_packet)));
+	Z_OOPS((req_copy.data = z_malloc(req->len))
+	       == NULL);
+
+	ret = z_impl_espi_read_request(dev, &req_copy);
+
+	Z_OOPS(z_user_to_copy(req, &req_copy,
+			      sizeof(struct espi_request_packet)));
+	Z_OOPS(z_user_to_copy(req->data, req_copy.data, req_copy.len));
+	k_free(req_copy.data);
+
+	return ret;
+}
+#include <syscalls/espi_read_request_mrsh.c>
+
+static inline int z_vrfy_espi_write_request(struct device *dev,
+					    struct espi_request_packet *req)
+{
+	int ret;
+	struct  espi_request_packet req_copy;
+
+	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, write_request));
+	Z_OOPS(z_user_from_copy(&req_copy, req,
+				sizeof(struct espi_request_packet)));
+	Z_OOPS((req_copy.data = z_user_alloc_from_copy(req->data, req->len))
```

Comment:
I'm not comfortable with all the heap allocations here and in other verification functions. Are they absolutely necessary?

Also any heap error should never Z_OOPS(), at most return -ENOMEM

Reply:
I was not sure about it too. The only thing doing that is that a lot of syscalls will may return ENOMEM now. Btw, I've asked it in the ticket you've opened.

Reply:
these buffers are user memory, these are the ones that we must copy. Otherwise it may happen of another thread change this inner buffer to point to a memory that the caller has no access and it will succeed.

Reply:
@andrew another problem in returning ENOMEM is how to propagate it in void syscalls

Reply:
> these buffers are user memory, these are the ones that we must copy. Otherwise it may happen of another thread change this inner buffer to point to a memory that the caller has no access and it will succeed.

this assumption is wrong, we have already copied the address to the local variable. But another thread still can tamper the buffer's contents.

Reply:
Oops I hit edit instead of replying.

Anyway, "but another thread still can tamper the buffer's contents." I don't see this as a problem. Can you expand if you disagree?

Reply:
From my understanding it is the same reason we have to make a copy of parameters passed by reference. Lets take this example:
```
static inline int z_vrfy_espi_config(struct device *dev, struct espi_cfg *cfg)
{
	struct  espi_cfg cfg_copy;

	Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, config));
	Z_OOPS(z_user_from_copy(&cfg_copy, cfg,
				sizeof(struct espi_cfg)));

	return z_impl_espi_config(dev, &cfg_copy);
}
```
Here we make a `cfg` copy because the implementation will use it to configuration and in this case if another thread change `cfg` values we will end up having a problem, that can be just a misconfiguration or something else. One function could use a given buffer to do the same kind of work, for instance one drive needs a blob to configure itself. But as you have  showed me, the policy is the implementation don't do logical  decisions based on given buffer, so we just need to ensure that everybody is following it (that I believe is the case)